### PR TITLE
8281104: jar --create should create missing parent directories

### PR DESCRIPTION
**Pull request by sormuras**
Calling `jar --create --file a/b/foo.jar INPUT-FILES` should create missing parent directories (here `a/b`) on the default file system before storing the JAR file (here `foo.jar`) in the destination directory.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281104](https://bugs.openjdk.java.net/browse/JDK-8281104): jar --create should create missing parent directories

### Reviewers
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7327/head:pull/7327` \
`$ git checkout pull/7327`

Update a local copy of the PR: \
`$ git checkout pull/7327` \
`$ git pull https://git.openjdk.java.net/jdk pull/7327/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7327`

View PR using the GUI difftool: \
`$ git pr show -t 7327`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7327.diff">https://git.openjdk.java.net/jdk/pull/7327.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jartool/share/classes/sun/tools/jar/Main.java [hunk 1]
```diff
@@ -461,7 +461,12 @@ private void validateAndClose(File tmpfile) throws IOException {
         try {
             if (ok) {
                 if (fname != null) {
-                    Files.move(path, Paths.get(fname), StandardCopyOption.REPLACE_EXISTING);
+                    Path target = Paths.get(fname);
+                    Path parent = target.getParent();
+                    if (parent != null) {
+                        Files.createDirectories(parent);
+                    }
```
Comment (LanceAndersen) [inline src/jdk.jartool/share/classes/sun/tools/jar/Main.java:468]: Would be good to move the creation after validating the arguments as this would fail after creating the temp zip file.  Be good to do via another PR

File: test/jdk/tools/jar/CreateMissingParentDirectories.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            doFailingTest(topDir.toString() + "/a/*/test.jar", entry);
+            Path blocker = Files.writeString(topDir.resolve("blocker.txt"), "Blocked!");
+            doFailingTest(topDir.resolve("blocker.txt/test.jar").toString(), entry);
+        } finally {
+            cleanup(topDir);
+        }
+    }
+
+    private static void doHappyPathTest(Path jar, Path entry) throws Throwable {
+        String[] jarArgs = new String[]{"cf", jar.toString(), entry.toString()};
```
Comment (LanceAndersen) [inline test/jdk/tools/jar/CreateMissingParentDirectories.java:78]: I might consider also using --create --file in addition to "cf" in a run for an extra sanity check

**Review by LanceAndersen — COMMENTED** [review]
Thank you for taking this on.  Overall looks good.

A few comments below.

Also, we should update jar.properties to indicate that the directory path will be created as needed in the help section for jar.

Best,
Lance

**Comment by sormuras** [comment]
Thanks for the review, Lance.

I didn't change order of creation, validation, and movement of the temporary JAR file in order to keep existing behaviour consistent.

A reason to use the main-based testing approach was also the ability to run the test via JEP 330:
`java CreateMissingParentDirectories.java`

> Also, we should update jar.properties to indicate that the directory path will be created as needed in the help section for jar.

That's a good point. Where would we put that information?
- [ ] Extend the descriptions of option `c`/`--create` and/or option `f`/`--file`?
- [ ] As a post scriptum note below the options?
  https://github.com/openjdk/jdk/blob/master/src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties#L309-L317

**Comment by LanceAndersen** [comment]
> Thanks for the review, Lance.
> 
> I didn't change order of creation, validation, and movement of the temporary JAR file in order to keep existing behaviour consistent.

I do think we are better served with the validation check earlier on.  In the case of a failure, we do not make the tmp file name known so it could be easy to miss. 

I think it makes sense to report the issue before erring out after creating the jar..  If the behavior stays as is, it would be best to document that the tmp file would still be there.

Let's see what others think.  Either way we can accomplish the additional change if we reach consensus via a separate PR.
> 
> A reason to use the main-based testing approach was also the ability to run the test via JEP 330: `java CreateMissingParentDirectories.java`

For me personally,  I would not use that means to run the tests as typically  would from an IDE.  Plus there is a benefit of leveraging the test framework in TestNG.    I totally understand everyone has a personal preference.  A discussion for another day I guess ;-).
> 
> > Also, we should update jar.properties to indicate that the directory path will be created as needed in the help section for jar.
> 
> That's a good point. Where would we put that information?
> 
> * [ ]  Extend the descriptions of option `c`/`--create` and/or option `f`/`--file`?

I think this would be the appropriate place for documenting the behavior.

> * [ ]  As a post scriptum note below the options?
>   https://github.com/openjdk/jdk/blob/master/src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties#L309-L317


Also we will need to update the MD file which represents the jar man page via a separate PR.

**Comment by sormuras** [comment]
> I think this would be the appropriate place for documenting the behavior.

Like this? 

```text
-c, --create               Create the archive including missing parent directories
...
-f, --file=FILE            The archive file name. When omitted, either stdin or
                           stdout is used based on the operation. Missing parent
                           directories of the file name path will be created
```

Perhaps, only adding it to `-c, --create` suffices. Having it also on `-f, --file` may confuse users, as this option is used all operation modes.

> Also we will need to update the MD file which represents the jar man page via a separate PR.

Yes. Will create PR for this.

**Comment by LanceAndersen** [comment]
> > I think this would be the appropriate place for documenting the behavior.
> 
> Like this?
> 
> ```
> -c, --create               Create the archive. When the path specified by -f, --file
>                            contains a path, missing parent directories will also be created
> ...
> -f, --file=FILE            The archive file name. When omitted, either stdin or
>                            stdout is used based on the operation. Missing parent
>                            directories of the file name path will be created
> ```
> 
> Perhaps, only adding it to `-c, --create` suffices. Having it also on `-f, --file` may confuse users, as this option is used all operation modes.

I think just having the verbiage when creating the jar should suffice as if we were updating it, the path would need to exist already.
> 
> > Also we will need to update the MD file which represents the jar man page via a separate PR.
> 
> Yes. Will create PR for this.

Great, thank you.  Have a good weekend!

**Comment by sormuras** [comment]
Should I also extend this "usage.compat" help message block? When and where is it displayed? 🤔 

```
usage.compat=\
\Compatibility Interface:\
\n\
Usage: jar {ctxui}[vfmn0PMe] [jar-file] [manifest-file] [entry-point] [-C dir] files] ...\n\
Options:\n\
\ \   -c  create new archive\n\
...
```

**Review by LanceAndersen — APPROVED** [review]
The current changes look good overall

**Comment by LanceAndersen** [comment]
> Should I also extend this "usage.compat" help message block? When and where is it displayed? 🤔
> 
> https://github.com/openjdk/jdk/blob/48523b090886f7b24ed4009f0c150efaa6f7b056/src/jdk.jartool/share/classes/sun/tools/jar/resources/jar.properties#L167-L173

I guess that makes sense.  Not sure how often anyone looks at this but good for consistency.  
Thanks for the reminder of the help message for compat

**Review by LanceAndersen — APPROVED** [review]
